### ISSUE TITLE
Adalite and Lilith on TPU

### ISSUE THREAD

**Issue by IsNoobgrammer** [euclaise/supertrainer2000#7]
When using Adalite and Lilith for the 2nd step of my training loop i got 
"bf16 loss[4000,4096] deleted on TPU:0"

however same hyperparams works for adam

**Comment by IsNoobgrammer** [comment]
![image](https://github.com/euclaise/supertrainer2000/assets/112808251/95145919-7ff2-4622-be75-3e6dde79f9b5)
![image](https://github.com/euclaise/supertrainer2000/assets/112808251/fdc93f44-eb62-4a1c-b8f9-3634c8ac64af)


this behaviour is only replicale with lilith and adalite optimizers and not with adam and other optimizers

**Comment by euclaise** [comment]
Does the Adafactor optimizer from [pytorch-optimizer](https://pytorch-optimizer.readthedocs.io/) work?

**Comment by IsNoobgrammer** [comment]
> Does the Adafactor optimizer from [pytorch-optimizer](https://pytorch-optimizer.readthedocs.io/) work?

yeah adafactor works correctly
![image](https://github.com/euclaise/supertrainer2000/assets/112808251/1b555551-28e3-47a9-9b6c-b6a40571ea37)



lilith
![image](https://github.com/euclaise/supertrainer2000/assets/112808251/cf61e3af-00a4-4af6-8b57-b3dc1cd6ab01)


everything is same except changing optimizer

**Comment by IsNoobgrammer** [comment]
```python

            outputs = model(input_ids=input_ids,attention_mask=attention_mask)
            loss = evaluate_loss(outputs,labels)
          
                
                
                
                
            del input_ids , attention_mask 
            loss.backward()
            del outputs,loss
            
            
            
            
            if (step+1) % FLAGS['GRAD_ACCUMULATION_STEP'] == 0:
                torch.nn.utils.clip_grad_norm_(update_params, max_norm=FLAGS['MAX_GRAD_CLIP']*8)
                scheduler.step()
                xm.optimizer_step(optimizer,pin_layout=True,barrier=True) ## performs xm.reduce_gradient() , optimizer.step() , 
                optimizer.zero_grad()
                
                
```

same script for both

**Comment by euclaise** [comment]
Try the current version, I just updated it to fix an issue with loss closures

**Comment by euclaise** [comment]
Let me know if the new version works

**Comment by IsNoobgrammer** [comment]
Since currently due to lack of resource i am not able to check it , i will close this and reopen when i will work , thanks
